File: java/src/main/java/com/example/client/ExampleClientRPC.java
package com.example.client;

import com.example.contract.PurchaseOrderState;
import com.google.common.net.HostAndPort;
import kotlin.Pair;
import net.corda.core.messaging.CordaRPCOps;
import net.corda.core.transactions.SignedTransaction;
import net.corda.node.services.config.ConfigUtilities;
import net.corda.node.services.messaging.CordaRPCClient;
import org.apache.activemq.artemis.api.core.ActiveMQException;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import rx.Observable;

import java.util.List;
import java.util.concurrent.ExecutionException;

/**
 * Demonstration of using the CordaRPCClient to connect to a Corda Node and
 * steam some State data from the node.
 */
public class ExampleClientRPC {
    public static void main(String[] args) throws ActiveMQException, InterruptedException, ExecutionException {
        if (args.length != 1) {
            throw new IllegalArgumentException("Usage: ExampleClientRPC <node address>");
        }

        final Logger logger = LoggerFactory.getLogger(ExampleClientRPC.class);
        final HostAndPort nodeAddress = HostAndPort.fromString(args[0]);
        final CordaRPCClient client = new CordaRPCClient(nodeAddress, ConfigUtilities.configureTestSSL());

        // Can be amended in the com.example.Main file.
        client.start("user1", "test");
        final CordaRPCOps proxy = client.proxy(null, 0);

        // Grab all signed transactions and all future signed transactions.
        final Pair<List<SignedTransaction>, Observable<SignedTransaction>> txsAndFutureTxs =
            proxy.verifiedTransactions();
        final List<SignedTransaction> txs = txsAndFutureTxs.getFirst();
        final Observable<SignedTransaction> futureTxs = txsAndFutureTxs.getSecond();

        // Log the 'placed' purchase order states and listen for new ones.
        futureTxs.startWith(txs).toBlocking().subscribe(
                transaction ->
                        transaction.getTx().getOutputs().forEach(
                                output -> {
                                    final PurchaseOrderState poState = (PurchaseOrderState) output.getData();
                                    logger.info(poState.getPurchaseOrder().toString());
                                })
        );
    }
}

File: src/main/kotlin/com/example/client/ExampleClientRPC.kt
package com.example.client

import com.google.common.net.HostAndPort
import net.corda.client.CordaRPCClient
import net.corda.core.transactions.SignedTransaction
import net.corda.core.utilities.loggerFor
import net.corda.node.services.config.configureTestSSL
import net.corda.node.services.messaging.CordaRPCOps
import org.graphstream.graph.Edge
import org.graphstream.graph.Node
import org.graphstream.graph.implementations.SingleGraph
import rx.Observable
import java.nio.file.Paths
import java.util.concurrent.CompletableFuture

/**
 *  Demonstration of using the CordaRPCClient to connect to a Corda Node and
 *  execute an exposed RPC command [CordaRPCOps]
 *
 *  Please follow instructions in the following tutorial to execute this example:
 *  https://docs.corda.r3cev.com/tutorial-clientrpc-api.html for further details.
 **/

class ExampleClientRPC {
}

private val log = loggerFor<ExampleClientRPC>()

enum class PrintOrVisualise {
    Print,
    Visualise
}

fun main(args: Array<String>) {

    require(args.size == 2) { "Usage: ExampleClientRPC <node address> [Print|Visualise]" }

    val nodeAddress = HostAndPort.fromString(args[0])
    val printOrVisualise = PrintOrVisualise.valueOf(args[1])

    val client = CordaRPCClient(nodeAddress, configureTestSSL())
    // Default user details from automatically created rpc-users.properties in the node folder.
    client.start("user1", "test")
    val proxy = client.proxy()

    val (transactions: List<SignedTransaction>, futureTransactions: Observable<SignedTransaction>) =
        proxy.verifiedTransactions()

    when (printOrVisualise) {
        PrintOrVisualise.Print -> {
            futureTransactions.startWith(transactions).subscribe { transaction ->
                log.info("NODE ${transaction.id}")
                transaction.tx.inputs.forEach { input ->
                    log.info("EDGE ${input.txhash} ${transaction.id}")
                }
            }
            CompletableFuture<Unit>().get()
        }
        PrintOrVisualise.Visualise -> {
            val graph = SingleGraph("transactions")
            transactions.forEach { transaction ->
                graph.addNode<Node>("${transaction.id}")
            }
            transactions.forEach { transaction ->
                transaction.tx.inputs.forEach { ref ->
                    graph.addEdge<Edge>("$ref", "${ref.txhash}", "${transaction.id}")
                }
            }
            futureTransactions.subscribe { transaction ->
                graph.addNode<Node>("${transaction.id}")
                transaction.tx.inputs.forEach { ref ->
                    graph.addEdge<Edge>("$ref", "${ref.txhash}", "${transaction.id}")
                }
            }
            graph.display()
        }
    }
}
